pkg/spdk/engine.go

```diff
@@ -1229,6 +1346,415 @@ func (e *Engine) validateAndUpdateReplicaNvme(replicaName string, bdev *spdktype
+		engineErr = errors.Wrap(err, "reconnectFrontend failed")
+		retErr = engineErr
+		return retErr
+	}
+	e.log.Info("Expanding engine complete")
+
+	expanded = true // which could be true even in partial success
+	return nil
+}
+
+func (e *Engine) startExpansion(size uint64, replicaClients map[string]*client.SPDKClient) (requireExpansion bool, err error) {
+	if e.isExpanding {
+		return false, fmt.Errorf("expansion is in progress")
+	}
+
+	if e.IsRestoring {
+		return false, fmt.Errorf("restoring is in progress")
+	}
+
+	if e.SpecSize > size {
+		return false, fmt.Errorf("cannot expand engine to a smaller size %v, current size %v", size, e.SpecSize)
+	} else if e.SpecSize == size {
+		e.log.Infof("Engine already at requested size %v, skipping expansion", size)
+		return false, nil // no need to expand
+	}
+
+	roundedNewSize := util.RoundUp(size, helpertypes.MiB)
+	if roundedNewSize != size {
+		return false, fmt.Errorf("rounded up spec size from %v to %v since the spec size should be multiple of MiB", size, roundedNewSize)
+	}
+
+	// Ensure all replicas are in RW mode and have the same size
+	if len(e.ReplicaStatusMap) == 0 {
+		e.log.Warn("Cannot expand engine: no replica found")
+		return false, fmt.Errorf("cannot expand engine with no replica")
+	}
+
+	currentReplicaSize := uint64(0)
+	for replicaName, replicaStatus := range e.ReplicaStatusMap {
+		if replicaStatus.Mode != types.ModeRW {
+			return false, fmt.Errorf("cannot expand engine with replica %s in mode %v", replicaName, replicaStatus.Mode)
+		}
+
+		replicaClient, ok := replicaClients[replicaName]
+		if !ok {
+			return false, fmt.Errorf("cannot find client for replica %s", replicaName)
+		}
+		replica, err := replicaClient.ReplicaGet(replicaName)
+		if err != nil {
+			return false, errors.Wrapf(err, "cannot get replica %s before expansion", replicaName)
+		}
+
+		if currentReplicaSize == 0 {
+			currentReplicaSize = replica.SpecSize
+		} else if currentReplicaSize != replica.SpecSize {
+			return false, fmt.Errorf("cannot expand engine with replicas in different sizes: replica %s has size %v while other replicas have size %v", replicaName, replica.SpecSize, currentReplicaSize)
+		}
+	}
+
+	if currentReplicaSize > size {
+		return false, fmt.Errorf("cannot expand engine to a smaller size %v, current replica size %v", size, currentReplicaSize)
+	} else if currentReplicaSize == size {
+		e.log.Infof("Replicas already at requested size %v, skipping expansion", size)
+		e.SpecSize = size
+		return false, nil // no need to expand
+	}
+
+	// Mark expansion as in progress
+	e.isExpanding = true
+	e.lastExpansionFailedAt = ""
+	e.lastExpansionError = ""
+
+	return true, nil
+}
+
+func (e *Engine) finishExpansion(expanded bool, size uint64) {
+	if expanded {
+		if e.lastExpansionError != "" {
+			e.log.Infof("Succeeded to expand from size %v to %v but there are some replica expansion failures: %v", e.SpecSize, size, e.lastExpansionError)
```

Comment:
If some relicas failed to expand, we still call it a successful expansion? What's the behavior in v1?

Reply:
We don't document the behavior of expansion failure in the official document.
@davidcheng0922 Can you create a DOC ticket to improve https://longhorn.io/docs/1.10.0/nodes-and-volumes/volumes/expansion/?